numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +153,178 @@ def potential_fn(z_gibbs, z_hmc):
+    proposal = random.randint(rng_proposal, (), minval=0, maxval=support_size - 1)
+    proposal = jnp.where(proposal >= z_discrete_flat[idx], proposal + 1, proposal)
+    z_discrete_new_flat = ops.index_update(z_discrete_flat, idx, proposal)
+    z_discrete_new = unravel_fn(z_discrete_new_flat)
+
+    pe_new = potential_fn(z_discrete_new)
+    accept_prob = jnp.clip(jnp.exp(pe - pe_new), a_max=1.0)
+    z_discrete, pe = cond(random.bernoulli(rng_accept, accept_prob),
+                          (z_discrete_new, pe_new), identity,
+                          (z_discrete, pe), identity)
+    return rng_key, z_discrete, pe
+
+
+def discrete_gibbs_fn(model, *model_args, **model_kwargs):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    Returns a gibbs_fn to be used in :class:`HMCGibbs`, which works for discrete latent sites
+    with enumerate support. Proposals are sampled uniformly over the set of possible values
+    in `enumerate_support` except for the current value (reference [1]).
+
+    Note that those discrete latent sites that are not specified in the constructor of
+    :class:`HMCGibbs` will be marginalized out by default (if they have enumerate supports).
+
+    **References**
+
+    1. *Peskun's theorem and a modified discrete-state Gibbs sampler*,
+       Liu, Jun S. (1996)
+
+    **Example**
+
+    .. doctest::
+
+        >>> from jax import random
+        >>> import jax.numpy as jnp
+        >>> import numpyro
+        >>> import numpyro.distributions as dist
+        >>> from numpyro.infer import MCMC, NUTS, HMCGibbs, discrete_gibbs_fn
+        ...
+        >>> def model(probs, locs):
+        ...     c = numpyro.sample("c", dist.Categorical(probs))
+        ...     numpyro.sample("x", dist.Normal(locs[c], 0.5))
+        ...
+        >>> probs = jnp.array([0.15, 0.3, 0.3, 0.25])
+        >>> locs = jnp.array([-2, 0, 2, 4])
+        >>> gibbs_fn = discrete_gibbs_fn(model, probs, locs)
+        >>> kernel = HMCGibbs(NUTS(model), gibbs_fn, gibbs_sites=["c"])
+        >>> mcmc = MCMC(kernel, 1000, 100000, progress_bar=False)
+        >>> mcmc.run(random.PRNGKey(0), probs, locs)
+        >>> samples = mcmc.get_samples()["x"]
+        >>> assert abs(jnp.mean(samples) - 1.3) < 0.1
+
+    """
+    # NB: all of the information such as `model`, `model_args`, `model_kwargs`
+    # can be accessed from HMCGibbs.sample but we require them here to
+    # simplify the api of `gibbs_fn`
+    prototype_trace = trace(seed(model, rng_seed=0)).get_trace(*model_args, **model_kwargs)
+    support_sizes = {
+        name: jnp.broadcast_to(site["fn"].enumerate_support(False).shape[0], jnp.shape(site["value"]))
+        for name, site in prototype_trace.items()
+        if site["type"] == "sample" and site["fn"].has_enumerate_support and not site["is_observed"]
+    }
+
+    def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
+        # convert to unconstrained values
+        z_hmc = {k: biject_to(prototype_trace[k]["fn"].support).inv(v)
```

Comment:
given our interface i guess this leads to a wasteful flow unconstrained => constrained => unconstrained?

Reply:
Yes, we also waste some other computations such as:
+ the initial potential_energy - this information iin HMCGibbs is not accessible through the signatures of `gibbs_fn`.
+ the last potential_energe - we recompute it in HMCGibbs